support/konnectivityproxy/dialer.go

```diff
@@ -0,0 +1,348 @@
+		client:                       opts.Client,
+		disableResolver:              opts.DisableResolver,
+		resolveFromGuestCluster:      opts.ResolveFromGuestClusterDNS,
+		resolveFromManagementCluster: opts.ResolveFromManagementClusterDNS,
+		dnsFallback:                  &proxy.fallbackToMCDNS,
+		log:                          opts.Log,
+	}
+	proxy.proxyResolver.guestClusterResolver = &guestClusterResolver{
+		client:               opts.Client,
+		konnectivityDialFunc: proxy.DialContext,
+		log:                  opts.Log,
+	}
+	return proxy, nil
+}
+
+type konnectivityProxy struct {
+	ca                              []byte
+	clientCert                      []byte
+	clientKey                       []byte
+	konnectivityHost                string
+	konnectivityPort                uint32
+	connectDirectlyToCloudAPIs      bool
+	resolveFromManagementClusterDNS bool
+	resolveBeforeDial               bool
+
+	proxyResolver
+
+	// fallbackToMCDNS is a synced boolean that keeps track
+	// of whether to fallback to the management cluster's DNS
+	// (and dial directly).
+	// It is initially false, but if lookup through the guest
+	// fails, then it's set to true.
+	fallbackToMCDNS syncBool
+
+	tlsConfigOnce sync.Once
+	tlsConfig     *tls.Config
+}
+
+func (p *konnectivityProxy) Dial(network, address string) (net.Conn, error) {
+	return p.DialContext(context.Background(), network, address)
+}
+
+func (p *konnectivityProxy) getTLSConfig() *tls.Config {
+	p.tlsConfigOnce.Do(func() {
+		certPool := x509.NewCertPool()
+		if !certPool.AppendCertsFromPEM(p.ca) {
+			panic("cannot load client CA")
+		}
+		p.tlsConfig = &tls.Config{
+			RootCAs:    certPool,
+			MinVersion: tls.VersionTLS12,
+		}
+		cert, err := tls.X509KeyPair(p.clientCert, p.clientKey)
+		if err != nil {
+			panic(fmt.Sprintf("cannot load client certs: %v", err))
+		}
+		p.tlsConfig.ServerName = p.konnectivityHost
+		p.tlsConfig.Certificates = []tls.Certificate{cert}
+	})
+	return p.tlsConfig
+}
+
+// dialFunc returns the appropriate dial function based on user and proxy setting configurations
+func (p *konnectivityProxy) DialContext(ctx context.Context, network string, requestAddress string) (net.Conn, error) {
+	requestHost, requestPort, err := net.SplitHostPort(requestAddress)
+	if err != nil {
+		return nil, fmt.Errorf("invalid address (%s): %w", requestAddress, err)
+	}
+	// return a dial direct function which respects any proxy environment settings
+	if p.connectDirectlyToCloudAPIs && isCloudAPI(requestHost) {
+		return p.dialDirectWithProxy(ctx, network, requestAddress)
+	}
+
+	// return a dial direct function ignoring any proxy environment settings
+	shouldDNSFallback := p.fallbackToMCDNS.get()
+	if shouldDNSFallback && p.resolveFromManagementClusterDNS {
+		return p.dialDirectWithoutProxy(ctx, network, requestAddress)
+	}
+
+	// get a TLS config based on x509 certs
+	tlsConfig := p.getTLSConfig()
+
+	// connect to the proxy address and get a TLS connection
```

Comment:
nit: I count at least 3 proxies in play: 
- Konnectivity-proxy -> The one outputted by this binary.
- The konnectivity server -> https://github.com/kubernetes-sigs/apiserver-network-proxy
- Https/http proxy -> system proxy configuration

The terminology proxy/konnectivity/httproxy becomes very confusing in this context. Just to be very specific, this could say "connect to konnectivity server address..." right?